peer-pod-controller/controllers/peerpod_controller.go

```diff
@@ -0,0 +1,148 @@
+	logger := log.FromContext(ctx)
+	pp := confidentialcontainersorgv1alpha1.PeerPod{}
+
+	if err := r.Get(ctx, req.NamespacedName, &pp); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if pp.ObjectMeta.GetDeletionTimestamp() == nil { // TODO: consider filter events
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.setProvider(ctx, pp.Spec.CloudProvider); err != nil { // TODO: set only once?
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("deleting instance", "InstanceID", pp.Spec.InstanceID, "CloudProvider", pp.Spec.CloudProvider)
+	if err := provider.DeleteInstance(ctx, pp.Spec.InstanceID); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if controllerutil.ContainsFinalizer(&pp, "peer.pod/finalizer") {
+		controllerutil.RemoveFinalizer(&pp, "peer.pod/finalizer")
+		if err := r.Update(ctx, &pp); err != nil {
+			if !apierrors.IsNotFound(err) { // object exist but fail to update, try again
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	logger.Info("instance deleted", "InstanceID", pp.Spec.InstanceID, "CloudProvider", pp.Spec.CloudProvider)
+	return ctrl.Result{}, nil
+}
+
+func getEnvOrDefault(key string, fallback ...string) string {
+	value := os.Getenv(key)
+	if len(value) == 0 && len(fallback) > 0 {
+		return fallback[0]
+	}
+	return value
+}
+
+// this mimics the variables setup in CAA, TODO: mount cm/secret directly?
+func (r *PeerPodReconciler) configsToEnv(cloudName string) {
+	peerpodscm := corev1.ConfigMap{}
+	peerpodssecret := corev1.Secret{}
+
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "peer-pods-cm-" + cloudName, Namespace: os.Getenv("PEERPOD_NAMESPACE")}, &peerpodscm)
+	if err == nil {
+		for k, v := range peerpodscm.Data {
+			os.Setenv(k, v)
+		}
+	}
+
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: "peer-pods-secret-" + cloudName, Namespace: os.Getenv("PEERPOD_NAMESPACE")}, &peerpodssecret)
+	if err == nil {
+		for k, v := range peerpodssecret.Data {
+			os.Setenv(k, string(v))
+		}
+	}
+}
+
+func (r *PeerPodReconciler) setProvider(ctx context.Context, cloudName string) error {
+	r.configsToEnv(cloudName)
+	switch cloudName {
+	case "aws":
+		var awscfg aws.Config
+		awscfg.AccessKeyId = getEnvOrDefault("AWS_ACCESS_KEY_ID")
+		awscfg.SecretKey = getEnvOrDefault("AWS_SECRET_ACCESS_KEY")
+		provider, _ = aws.NewProvider(&awscfg)
+	case "libvirt":
+		var libvirtcfg libvirt.Config
+		libvirtcfg.URI = getEnvOrDefault("LIBVIRT_URI")
+		libvirtcfg.PoolName = getEnvOrDefault("LIBVIRT_POOL", "default")
+		libvirtcfg.NetworkName = getEnvOrDefault("LIBVIRT_NET", "default")
+		provider, _ = libvirt.NewProvider(&libvirtcfg)
+	default:
+		return fmt.Errorf("unknown provider: %s", cloudName)
+	}
+
+	if cloud := cloudmgr.Get(cloudName); cloud != nil {
+		cloud.LoadEnv() // this doesn't work ATM as it requires providers build tags for cloudmgr
+	}
```

Comment:
If we can specify build tags here, I think we can simplify these lines to something like this.

```
	cloud := cloudmgr.Get(cloudName)
	if cloud == nil {
		return fmt.Errorf("unknown provider: %s", cloudName)
	}
	provider, err := cloud.NewProvider()
	if err != nil {
		return err
	}
```

Reply:
I couldn't find how to do it, IIRC adding tags caused the controller-gen to fail
I thought we may be able to go the other way around, i.e. to build all unless there's a tag

Reply:
done